cmd/snap/cmd_components.go

```diff
@@ -0,0 +1,135 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/snap/naming"
+)
+
+var shortComponentsHelp = i18n.G("List available and installed components for installed snaps")
+var longComponentsHelp = i18n.G(`
+The components command displays a summary of the components that are installed
+and available for the set of currently installed snaps.
+
+Components for specific installed snaps can be queried by providing snap names
+as positional arguments.
+`)
+
+type cmdComponents struct {
+	clientMixin
+	Positional struct {
+		Snaps []installedSnapName `positional-arg-name:"<snap>"`
+	} `positional-args:"yes"`
+}
+
+func init() {
+	addCommand("components",
+		shortComponentsHelp,
+		longComponentsHelp,
+		func() flags.Commander { return &cmdComponents{} },
+		nil,
+		[]argDesc{{
+			name: i18n.G("<snap>"),
+			desc: i18n.G("Snaps to consider when listing available and installed components."),
+		}},
+	)
+}
+
+func (x *cmdComponents) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	names := installedSnapNames(x.Positional.Snaps)
+	snaps, err := x.client.List(names, nil)
+	if err != nil {
+		if err == client.ErrNoSnapsInstalled {
+			if len(names) == 0 {
+				fmt.Fprintln(Stderr, i18n.G("No snaps are installed yet. Try 'snap install hello-world'."))
+				return nil
+			} else {
+				return ErrNoMatchingSnaps
+			}
+		}
+		return err
+	}
+
+	anyComps := false
+	for _, snap := range snaps {
+		anyComps = anyComps || len(snap.Components) > 0
+	}
+
+	if !anyComps {
+		fmt.Fprintln(Stderr, i18n.G("No components are available for any installed snaps."))
+		return nil
+	}
+
+	sort.Sort(snapsByName(snaps))
+
+	w := tabWriter()
+	fmt.Fprintln(w, i18n.G("Component\tStatus\tType"))
+	for _, snap := range snaps {
+		for _, comp := range sortedComponents(snap.Components) {
```

Comment:
FWIW, I wouldn't bother too much about making a copy of the slice, just sort it in place.

```suggestion
	for _, snap := range snaps {
		slice.Sort(componentsByInstallStatusAndSnapName(snap.Components)
		for _, comp := snap.Components {
```